app/src/command.rs

```diff
@@ -191,7 +215,25 @@ impl<'app, D: Dependencies<'app>> Application<'app, D> {
             );
             UnrecoverableAppFailure
         })?;
-        in_progress_payloads.finish_id(block, transactions.into_iter().map(Into::into));
+
+        // Mempool should only be cleaned after all other building stages succeeded
+        if !no_tx_pool {
+            // Payload transactions always come first in the block
+            new_transactions
+                .iter()
+                .skip(attributes_txs_len)
+                .for_each(|tx| {
+                    self.mem_pool.remove_by_hash(tx.tx_hash(), tx.signer());
```

Comment:
I just noticed that it would be more efficient to remove from the mempool by signer + nonce instead of signer  + hash because then we do not need to iterate through all the transactions from that sender looking for the right hash. We have access to the nonce here so I think it makes more sense to do that.

Reply:
Good catch! I should have considered that from the start, now that I think about it. Only had to add some failure conditions to tx casting, but it's even more explicit that way.